interop-testing/src/main/java/io/grpc/testing/istio/EchoTestServer.java

```diff
@@ -0,0 +1,368 @@
+      }
+    }
+  }
+
+  private static class EchoTestServerInterceptor implements ServerInterceptor {
+
+    @Override
+    public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call,
+        final Metadata requestHeaders, ServerCallHandler<ReqT, RespT> next) {
+      final String methodName = call.getMethodDescriptor().getBareMethodName();
+
+      // we need this processing only for Echo
+      if (!"Echo".equals(methodName)) {
+        return next.startCall(call, requestHeaders);
+      }
+
+      final SocketAddress peerAddress = call.getAttributes()
+          .get(Grpc.TRANSPORT_ATTR_REMOTE_ADDR);
+
+      return next.startCall(
+          new ForwardingServerCall.SimpleForwardingServerCall<ReqT, RespT>(call) {
+
+            @SuppressWarnings("unchecked")
+            @Override
+            public void sendMessage(RespT message) {
+              io.grpc.testing.istio.Istio.EchoResponse echoResponse =
+                  (io.grpc.testing.istio.Istio.EchoResponse) message;
+              String oldMessage = echoResponse.getMessage();
+
+              EchoMessage echoMessage = new EchoMessage();
+
+              for (String key : requestHeaders.keys()) {
+                if (!key.endsWith("-bin")) {
+
+                  echoMessage.writeKeyValueForRequest(REQUEST_HEADER, key,
+                      requestHeaders.get(Metadata.Key.of(key, Metadata.ASCII_STRING_MARSHALLER)));
+                }
+              }
+              // This is not a complete list. May need to add/remove fields later,
+              // such as "ServiceVersion", "ServicePort", "URL", "Method", "ResponseHeader",
+              // "Cluster", "IstioVersion"
+              // Only keep the fields needed for now.
+              if (peerAddress instanceof InetSocketAddress) {
+                InetSocketAddress inetPeerAddress = (InetSocketAddress) peerAddress;
+                echoMessage.writeKeyValue(IP, inetPeerAddress.getAddress().getHostAddress());
+              }
+              echoMessage.writeKeyValue(STATUS_CODE, "200");
+              echoMessage.writeKeyValue(HOST, call.getAuthority());
+              echoMessage.writeMessage(oldMessage);
+              echoResponse =
+                  io.grpc.testing.istio.Istio.EchoResponse.newBuilder()
+                      .setMessage(echoMessage.toString())
+                      .build();
+              super.sendMessage((RespT) echoResponse);
+            }
+          },
+          requestHeaders);
+    }
+  }
+
+  private static class EchoTestServiceImpl extends EchoTestServiceImplBase {
+
+    private final String hostname;
+
+    EchoTestServiceImpl(String hostname) {
+      this.hostname = hostname;
+    }
+
+    @Override
+    public void echo(io.grpc.testing.istio.Istio.EchoRequest request,
+        io.grpc.stub.StreamObserver<io.grpc.testing.istio.Istio.EchoResponse> responseObserver) {
+
+      StringBuilder sb = new StringBuilder();
```

Comment:
This can be replaced with the EchoMessage as well, right?

Reply:
Good point. Done